### PR TITLE
Fix #1503: anonymous application of FunctionX

### LINKED ISSUE #1503: Anonymous functions direct application crash the compiler.
```scala
object Foo {
  (new Function0[Unit] {
    def apply() = println("hello")
  })()
}
```

Note that this does not fail if the `apply` method is written explicitly.

```
Exception in thread "main" java.lang.Error: unexpected type.
 fun = Block(List(TypeDef($anon,Template(DefDef(<init>,List(),List(List()),TypeTree[TypeRef(ThisType(TypeRef(NoPrefix,scala)),Unit)],EmptyTree),List(Apply(Select(New(TypeTree[TypeRef(ThisType(TypeRef(NoPrefix,lang)),Object)]),<init>),List()), AppliedTypeTree(Ident(Function0),List(Ident(Unit)))),ValDef(_,EmptyTree,EmptyTree),List(DefDef(apply,List(),List(List()),TypeTree[TypeRef(ThisType(TypeRef(NoPrefix,$anon)),scala$Function0$$R)],Apply(Ident(println),List(Literal(Constant(hello))))))))),Typed(Select(Apply(Select(New(Ident($anon)),<init>),List()),apply),TypeTree[MethodType(List(), List(), TypeRef(TermRef(ThisType(TypeRef(NoPrefix,<root>)),scala),Unit))])),
 methPart(fun) = Typed(Select(Apply(Select(New(Ident($anon)),<init>),List()),apply),TypeTree[MethodType(List(), List(), TypeRef(TermRef(ThisType(TypeRef(NoPrefix,<root>)),scala),Unit))]),
 methPart(fun).tpe = MethodType(List(), List(), TypeRef(TermRef(ThisType(TypeRef(NoPrefix,<root>)),scala),Unit)),
 tpe = MethodType(List(), List(), TypeRef(TermRef(ThisType(TypeRef(NoPrefix,<root>)),scala),Unit))
	at dotty.tools.dotc.typer.Applications$class.handleUnexpectedFunType(Applications.scala:661)
	at dotty.tools.dotc.typer.Typer.handleUnexpectedFunType(Typer.scala:60)
	at dotty.tools.dotc.typer.Applications$$anonfun$realApply$1$1.dotty$tools$dotc$typer$Applications$class$$anonfun$$simpleApply$1(Applications.scala:577)
	at dotty.tools.dotc.typer.Applications$$anonfun$realApply$1$1$$anonfun$apply$5.apply(Applications.scala:603)
	at dotty.tools.dotc.typer.Applications$$anonfun$realApply$1$1$$anonfun$apply$5.apply(Applications.scala:603)
	at dotty.tools.dotc.typer.Typer.tryEither(Typer.scala:1472)
	at dotty.tools.dotc.typer.Applications$$anonfun$realApply$1$1.apply(Applications.scala:604)
	at dotty.tools.dotc.typer.Applications$$anonfun$realApply$1$1.apply(Applications.scala:544)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:36)
	at dotty.tools.dotc.typer.Applications$class.realApply$1(Applications.scala:544)
	at dotty.tools.dotc.typer.Applications$class.typedApply(Applications.scala:647)
	at dotty.tools.dotc.typer.Typer.typedApply(Typer.scala:60)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1376)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1417)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1427)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1425)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:155)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:56)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1425)
	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:1455)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1460)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedClassDef$1.apply(Typer.scala:1196)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedClassDef$1.apply(Typer.scala:1144)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:36)
	at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:1144)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1367)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1415)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1427)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1425)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:155)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:56)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1425)
	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:1449)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1460)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedPackageDef$1.apply(Typer.scala:1292)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedPackageDef$1.apply(Typer.scala:1283)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:36)
	at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:1283)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1405)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1417)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1427)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1425)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:155)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:56)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1425)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1464)
	at dotty.tools.dotc.typer.FrontEnd$$anonfun$typeCheck$1.apply$mcV$sp(FrontEnd.scala:47)
	at dotty.tools.dotc.typer.FrontEnd.monitor(FrontEnd.scala:23)
	at dotty.tools.dotc.typer.FrontEnd.typeCheck(FrontEnd.scala:45)
	at dotty.tools.dotc.typer.FrontEnd$$anonfun$runOn$3.apply(FrontEnd.scala:69)
	at dotty.tools.dotc.typer.FrontEnd$$anonfun$runOn$3.apply(FrontEnd.scala:69)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at dotty.tools.dotc.typer.FrontEnd.runOn(FrontEnd.scala:69)
	at dotty.tools.dotc.Run$$anonfun$compileUnits$1$$anonfun$apply$mcV$sp$1.apply(Run.scala:69)
	at dotty.tools.dotc.Run$$anonfun$compileUnits$1$$anonfun$apply$mcV$sp$1.apply(Run.scala:66)
	at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
	at dotty.tools.dotc.Run$$anonfun$compileUnits$1.apply$mcV$sp(Run.scala:66)
	at dotty.tools.dotc.Run$$anonfun$compileUnits$1.apply(Run.scala:60)
	at dotty.tools.dotc.Run$$anonfun$compileUnits$1.apply(Run.scala:60)
	at dotty.tools.dotc.util.Stats$.monitorHeartBeat(Stats.scala:69)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:60)
	at dotty.tools.dotc.Run.compileSources(Run.scala:57)
	at dotty.tools.dotc.Run.compile(Run.scala:41)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:26)
	at dotty.tools.dotc.Driver.process(Driver.scala:124)
	at dotty.tools.dotc.Driver.process(Driver.scala:93)
	at dotty.tools.dotc.Driver.process(Driver.scala:105)
	at dotty.tools.dotc.Driver.main(Driver.scala:132)
	at dotty.tools.dotc.Main.main(Main.scala)
```

### PR DESCRIPTION
**Pull request by felixmulder**

### CODE REVIEW

File: src/dotty/tools/dotc/typer/Applications.scala [hunk 1]
```diff
@@ -573,6 +573,12 @@ trait Applications extends Compatibility { self: Typer with Dynamic =>
               else
                 new ApplyToUntyped(tree, fun1, funRef, proto, pt)(argCtx(tree))
             convertNewGenericArray(ConstFold(app.result))
+          case mt: MethodType =>
```
Comment (smarter) [inline src/dotty/tools/dotc/typer/Applications.scala]: I would squash this commit and the next one in the third commit, no need to record everything that was tried.

File: src/dotty/tools/dotc/typer/Applications.scala [hunk 2]
```diff
@@ -607,16 +615,16 @@ trait Applications extends Compatibility { self: Typer with Dynamic =>
...
-            (failedVal, failedState) =>
-              def fail = { failedState.commit(); failedVal }
-              // Try once with original prototype and once (if different) with tupled one.
-              // The reason we need to try both is that the decision whether to use tupled
-              // or not was already taken but might have to be revised when an implicit
-              // is inserted on the qualifier.
-              tryWithImplicitOnQualifier(fun1, originalProto).getOrElse(
-                if (proto eq originalProto) fail
-                else tryWithImplicitOnQualifier(fun1, proto).getOrElse(fail))
+          } { (failedVal, failedState) =>
```
Comment (smarter) [inline src/dotty/tools/dotc/typer/Applications.scala]: style changes should be in separate commits from functional changes

File: tests/run/i1503.scala [hunk 3]
```diff
@@ -0,0 +1,7 @@
+object Test {
```
Comment (smarter) [inline tests/run/i1503.scala]: I would put the run test in the same commit where the problem is fixed

File: tests/pos/i1503.scala [hunk 4]
```diff
@@ -0,0 +1,6 @@
+object Test {
+  (
+new Function0[Unit] {
+    def apply() = println("hello")
```
Comment (smarter) [inline tests/pos/i1503.scala]: Do we really need a separate pos test since the same code will be compiled by the run test?

File: src/dotty/tools/dotc/typer/Typer.scala [hunk 5]
```diff
@@ -533,7 +533,23 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
...
+     *  {{{
+     *  (new Function0[Unit] {
+     *    def apply(): Unit = ...
+     *  })()
+     *  }}}
+     *
+     *  The code gets desugared into a block that is applied. In the case
+     *  above, the block ends with a `Select(_ apply)` in expression position
+     *  and fails compilation. To remedy this, if a block ends in a
+     *  `Select(_,apply)` we simply move the apply outside the block.
```
Comment (smarter) [inline src/dotty/tools/dotc/typer/Typer.scala]: This comment does not make it clear that the apply does not come from the desugaring but from the adaptation,  and that adaptation will add the .apply back around the block, here's my attempt at explaining what's going ont:
```scala
    // A block might appear in function position after desugaring (see i1503.scala),
    // so adaptation may insert a `.apply` at the end of expr1. When this happens we
    // simply remove the `.apply` to make expr1 a valid block expression, the `.apply`
    // will be added back around the block when the block itself is adapted.
```

File: tests/run/i1503.scala (hunk 3 above)
Comment (felixmulder) [inline tests/run/i1503.scala]: Agreed, was planning on cleaning up this PR this morning - thanks for having a look already :)

File: src/dotty/tools/dotc/typer/Applications.scala (hunk 2 above)
Comment (felixmulder) [inline src/dotty/tools/dotc/typer/Applications.scala]: Done 👍

**Comment by felixmulder** [comment]
Addressed @smarter's concerns - we'd like input by @odersky on the fix however, so waiting on that before merging.

File: src/dotty/tools/dotc/typer/Typer.scala [hunk 6]
```diff
@@ -533,7 +533,16 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
...
     val exprCtx = index(tree.stats)
     val stats1 = typedStats(tree.stats, ctx.owner)
-    val expr1 = typedExpr(tree.expr, pt)(exprCtx)
+
+    // A block might appear in function position after desugaring (see
+    // i1503.scala), so adaptation may insert a `.apply` at the end of expr1.
+    // When this happens we simply remove the select to make expr1 a valid
+    // block expression. The `.apply` will be added back around the block when
+    // the block itself is adapted later.
+    val expr1 = typedExpr(tree.expr, pt)(exprCtx) match {
```
Comment (odersky) [inline src/dotty/tools/dotc/typer/Typer.scala]: But what if the .apply was added explicitly by the user? Then it should not be removed. It seems we need another solution.

**Comment by odersky** [comment]
Superseded by #1522.
